pkg/agent/git/service.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2024 Daytona Platforms Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package git
+
+import (
+	"bytes"
+	"os"
+	"path"
+
+	"github.com/daytonaio/daytona/pkg/agent/config"
+	"github.com/daytonaio/daytona/pkg/serverapiclient"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"gopkg.in/ini.v1"
+)
+
+type Service struct {
+	Config *config.Config
+}
+
+func (s *Service) CloneRepository(project *serverapiclient.Project, authToken *string) error {
+	cloneOptions := &git.CloneOptions{
+		URL:             *project.Repository.Url,
+		Progress:        os.Stdout,
+		SingleBranch:    true,
+		InsecureSkipTLS: true,
+	}
+
+	if authToken != nil {
+		cloneOptions.Auth = &http.BasicAuth{
+			Username: "daytona",
+			Password: *authToken,
+		}
+	}
+
+	if s.shouldCloneBranch(project) {
+		cloneOptions.ReferenceName = plumbing.ReferenceName("refs/heads/" + *project.Repository.Branch)
+	}
+
+	_, err := git.PlainClone(s.Config.ProjectDir, false, cloneOptions)
+	if err != nil {
+		return err
+	}
+
+	if s.shouldCheckoutSha(project) {
+		repo, err := git.PlainOpen(s.Config.ProjectDir)
+		if err != nil {
+			return err
+		}
+
+		w, err := repo.Worktree()
+		if err != nil {
+			return err
+		}
+
+		err = w.Checkout(&git.CheckoutOptions{
+			Hash: plumbing.NewHash(*project.Repository.Sha),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func (s *Service) RepositoryExists(project *serverapiclient.Project) (bool, error) {
+	_, err := os.Stat(s.Config.ProjectDir)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *Service) SetGitConfig(userData *serverapiclient.GitUserData) error {
+	gitConfigFileName := path.Join(os.Getenv("HOME"), ".gitconfig")
```

Comment:
This should also be a param of the service. It this was ran as a test, it would modify the gitconfig of the user running the test.